buffer/src/main/java/io/netty5/buffer/internal/CleanerPool.java

```diff
@@ -0,0 +1,155 @@
+
+    /**
+     * Size of the shared Cleaner pool used by external threads, and optionally by event-loop threads.
+     * @see #CNF_POOL_SIZE
+     */
+    private static final int POOL_SIZE = getPoolSize(1);
+
+    /**
+     * Flag to configure whether event-loop threads must use the shared cleaner pool (default=false).
+     * @see #CNF_EVENT_LOOP_USE_POOL
+     */
+    private static final boolean EVENT_LOOP_USE_POOL = SystemPropertyUtil.getBoolean(CNF_EVENT_LOOP_USE_POOL, false);
+
+    /**
+     * Cleaner instances shared by all external threads, and optionally by event-loop threads
+     * These cleaners are wrapped in a static inner class for lazy initialization purpose.
+     */
+    private static class CleanersPool {
+        static final Cleaner[] cleaners = IntStream.range(0, POOL_SIZE)
+                .mapToObj(i -> createCleaner()).toArray(Cleaner[]::new);
+    }
+
+    /**
+     * A FastThreadLocal that returns a Cleaner to the caller thread.
+     * For event-loop threads, by default a dedicated cleaner instance is allocated and mapped to the
+     * caller thread, but by configuration, event-loop threads can use the shared cleaner pool if this is
+     * necessary.
+     * For external threads, a cleaner is always returned from the shared fixed cleaner pool in a round-robin way.
+     */
+    private final CleanerThreadLocal threadLocalCleaner = new CleanerThreadLocal();
+
+    /**
+     * The singleton for the CleanerPool.
+     */
+    public static final CleanerPool INSTANCE = new CleanerPool();
+
+    /**
+     * The FastThreadLocal used to map a Cleaner instance to Event Loop threads.
+     */
+    private static class CleanerThreadLocal extends FastThreadLocal<Cleaner> {
+        private static final AtomicInteger counter = new AtomicInteger();
+
+        @Override
+        protected Cleaner initialValue() {
+            if (!EVENT_LOOP_USE_POOL && ThreadExecutorMap.currentExecutor() != null) {
+                // Allocate one dedicated cleaner for the caller event-loop thread
+                return createCleaner();
+            }
+
+            // Return one of the shared cleaners from the shared cleaner pool.
+            return CleanersPool.cleaners[(counter.getAndIncrement() & 0x7F_FF_FF_FF) % POOL_SIZE];
+        }
+    }
+
+    private CleanerPool() {
+        logger.debug("Instantiating CleanerPool: {}={}, {}={}",
+                CNF_EVENT_LOOP_USE_POOL, EVENT_LOOP_USE_POOL,
+                CNF_POOL_SIZE, POOL_SIZE);
+    }
+
+    /**
+     * Returns a Cleaner to the calling thread.
+     * The same thread will get the same cleaner for every getCleaner method calls.
+     */
+    Cleaner getCleaner() {
+        return threadLocalCleaner.get();
+    }
+
+    /**
+     * Creates a Cleaner instance.
+     */
+    private static Cleaner createCleaner() {
+        if (logger.isDebugEnabled()) {
+            logger.debug("Creating cleaner.");
```

Comment:
I think we should remove the logging here and better do the logging in `initialValue` where we can log if we created a new one or used one from the pool. This seems more useful.

Reply:
done, logging is now performed in initialValue, as you suggested.